### PR TITLE
Create filters_agressive

### PR DESCRIPTION
**Pull request by NiJoao**
This is my tentative filters file.
My goal is a set-it-and-forget-it, to deploy in my families computers, consuming the least ammount of storage from my FTP server, and when something bad happens I must be able to recover everything from this backup and the internet.

I removed the MacOS part because I really cannot help there.
But feel free to merge it or use it as you like, and I'll be watching for upgrades on the main file ;)

### CODE REVIEW

File: filters_agressive [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+
+
+# ==============================================================================
+# ==============================================================================
+# ============================================================================== generic excludes (win and mac)
+# ==============================================================================
+# ==============================================================================
+
+# ============================ exclude the ".git" subfolder in repo root. If you got this folder from git
+e:(?i)^\.git\/.*
```
Comment (TheBestPessimist) [inline filters_agressive]: q: this does not exclude `(...)\somefolder\a repo\.git\`, but only `.git\` (as in `.git\` is in the repository root). Is this wanted?

**Comment by NiJoao** [comment]
At this point, yes.
Although I only left it because you had something similar in the original.
And since this file/scripts are in a git repo, it makes some sense.

Personally, I removed that line from my version. I think we should remove
all or none versioned folders, doesn't matter if it's from this repo or
other.
Later on the file I exclude all .git and .svn folders, since these are
heavy (large number of files) and easily replicated/recoverable.
But maybe we should keep the folder and remove the contents, to know
post-mortem it was versioned.

What's your opinion?

**Comment by TheBestPessimist** [comment]
Looking more and more at your file, i think that i like it better if each exclude is on its own line instead of having them merged.

For example:
i like better 
```
e:(?i)\$RECYCLE\.BIN/
e:(?i)\$RECYCLED/
e:(?i)\$RECYCLER/
```
instead of
```
e:(?i)^[^\/]+\/(\$RECYCLE\.BIN|\$RECYCLED|\$RECYCLER)\/.*
```

**Comment by TheBestPessimist** [comment]
In this case, what i will do is: take stuff from your file and add them to mine. 
Is this okay with you?

File: filters_agressive [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+
+# ============================ exclude recycle bins
+e:(?i)^[^\/]+\/(\$RECYCLE\.BIN|\$RECYCLED|\$RECYCLER)\/.*
+
+# ============================ THE windows folder should be excluded
+e:(?i)^[^\/]+\/Windows(\.old)?\/.*
+
+
+# ============================ system folders
+e:(?i)^[^\/]+\/(\$GetCurrent|\$SysReset|Boot|Config.Msi|Documents and Settings|MSOCache|Recovery|System Volume Information|Microsoft\/Windows)\/.*
```
Comment (TheBestPessimist) [inline filters_agressive]: What is $Getcurrent and $SysReset ? I don't remember seeing files named like that on my systems.
Is this some sort of variable replacement like String.format("${lalalulu} is awesome", "George")?

**Comment by NiJoao** [comment]
Hi,
That is fine by me, take whatever you want.

Regarding the $GetCurrent, they really are left-over windows folders,
present in some of mine:
https://www.howtogeek.com/308301/what-are-the-getcurrent-and-sysreset-folders-and-can-you-delete-them/


I've put that filter file in production use, and it is compacting the
backups enough for a C: use.
I'm very happy with it.
However, I'm having problems with Duplicacy itself, making the PCs very
slow while scanning and backing up (2/3min each hour).
I'm changing your other files to create a backup task using Background
priority.

Regarding the one-line-per-rule, it's really subjective.
I follow the rule:
Only do regex ORs with simple file/folder names, and never multi-level
folders and between different types.
Knowing this, it's somewhat easy to read: From the beginning and end you
know where are the exclusions and what type (e.g. root or User's folder;
files by name, extension, or folders?)
And everything inside the OR is the multiple files/folders/extensions to
exclude.
I find it a good compromise of compactness, performance and readability,
but feel free to do whatever you want with it ;)

Regards

File: filters_agressive [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+
+# ============================ useless files anywhere
+e:(?i).*(desktop\.ini|thumbs\.db)$
+
+# ============================ useless files anywhere, by extension
+e:(?i).*\.(te?mp|cache|crdownload|!ut|part|rbf|chk|dov|obj|bac|bak|bkp|bup|back|dmp|dump|err|old|prv)$
+
+
+# ============================ exclude program files, but keep the folder names to use as a list of installed apps
+e:(?i)^[^\/]+\/Programa?s(\.old)?\/[^\/]+\/.+
```
Comment (TheBestPessimist) [inline filters_agressive]: what's `Programa?s`?

**Comment by NiJoao** [comment]
Ho, that is a English-Portuguese mixed line (matching Programs AND
Programas).
Sorry, language specific (Programas) should be in a separate line...
Leftovers from ancient versions of my regexs. :(

**Comment by TheBestPessimist** [comment]
If you have any other updates, feel free to update this pull request. I am working on merging yours atm.
I still have a lot to learn about regex, so thanks for your examples :)

**Comment by TheBestPessimist** [comment]
Also, thanks for letting me merge your stuff into mine!

**Comment by TheBestPessimist** [comment]
Finally finished the merge. 

Wanna check out the filters file? There are a few modifications compared to yours though.

**Comment by NiJoao** [comment]
Hi!
The merging is almost perfect!
I have a few suggestions though:

- The note 3) should be number 1), as it is the most important one. And I would even put a "summary" on the top of the file, stating the filter file is_designed_to/only works in the "<symlinking backups> mode" or "Symlink_to_root mode.

- There are a few cases of different policies or not yet corrected. Such as:
```
e:/ProgramData/Microsoft/Network/Downloader/
e:ProgramData/Microsoft/Windows/WER/
e:(?i)\$Windows\.\~BT/
e:(?i)^[^/]*Windows10Upgrade/
```
which following the rules of the remaining file, with <case-insensitiveness> and <Root-only check>, etc, should be:
```
# The first 2 are also excluded later in the file by e:(?i)^[^/]+/ProgramData/Microsoft[^/]*/, but we can reinforce their uselessness here
e:(?i)^[^/]+/ProgramData/Microsoft/Network/Downloader/
e:(?i)^[^/]+/ProgramData/Microsoft/Windows/WER/
e:(?i)^[^/]+/\$Windows\.\~BT/
e:(?i)^[^/]+/\$Windows\.\~WS/
e:(?i)^[^/]+/Windows10Upgrade/
```
Are you still working on these cases, or can I make a new pull request?
I'll adopt your filter file, fix these, and add some language-specific rules of my own

3) Didn't you like the idea of this?
```
# ============================ exclude program files, but keep the folder names to use as a list of installed apps
# e:(?i)^[^/]+/Program Files[^/]*/[^/]+/.+
```
instead of ?
`e:(?i)^[^/]+/Program Files[^/]*/`

4) What's the problem with the Thunderbird profile include directive? 
```
# ============================ include Thunderbird email profile but nothing else
# note: may be wrong!
# i:(?i)/AppData/[^/]+/Thunderbird/$
# i:(?i)/AppData/[^/]+/Thunderbird/Profiles/
# e:(?i)/AppData/[^/]+/Thunderbird/
```
Paths taken from http://kb.mozillazine.org/Profile_folder_-_Thunderbird
The Thunderbird folder contains a bunch of other useless files, but I can accept if you consider it safer to backup everything in there.

**Comment by TheBestPessimist** [comment]
### Other rules

Fixed.

### Program files regex:
```
# ============================ exclude program files, but keep the folder names to use as a list of installed apps
# e:(?i)^[^/]+/Program Files[^/]*/[^/]+/.+
```
Because for me it doesn't work properly:

it matches all of these:
```
c__link/Program Files (x86)/Corsair/
c__link/Program Files (x86)/Corsair/Corsair Utility Engine/
c__link/Program Files (x86)/Corsair/Corsair Utility Engine/audio/
c__link/Program Files (x86)/Corsair/Corsair Utility Engine/devices/
c__link/Program Files (x86)/Corsair/Corsair Utility Engine/devices/masks/
c__link/Program Files (x86)/Corsair/Corsair Utility Engine/devices/masks/dark_core_rgb/
c__link/Program Files (x86)/Corsair/Corsair Utility Engine/devices/masks/glaive_rgb/
c__link/Program Files (x86)/Corsair/Corsair Utility Engine/devices/masks/harpoon_rgb/
c__link/Program Files (x86)/Corsair/Corsair Utility Engine/devices/masks/hs60surround/
c__link/Program Files (x86)/Corsair/Corsair Utility Engine/devices/masks/hs70wireless/
c__link/Program Files (x86)/Corsair/Corsair Utility Engine/devices/masks/k55/
```

which goes waaay over the "keep folder names at the level just after progra~1"
(it doesn't go just 1 level deep in progra~1, but all the way).

I intend to add it evetually, at least as a commented option -- as it is useful, but first: it should work as intended.
(therefore, if you already have it fixed, send me a pull request, put it here in a comment, i don't know, and i'll take it :^) )


### What's the problem with the Thunderbird profile include directive?

The problem with thunderbird is that i cant check it, but i believe it has the same problem i had with firefox: not all the data is saved.

My firefox rules:
```
# ============================ include Firefox profile but nothing else from Mozilla
i:(?i)/AppData/[^/]+/Mozilla/$
i:(?i)/AppData/[^/]+/Mozilla/Firefox/
e:(?i)/AppData/[^/]+/Mozilla/
```

they also include the file: `AppData\Roaming\Mozilla\Firefox\profiles.ini` which is needed to have a working firefox state (it has the "current profile name" and something else in it).

Eith the filter you provided:
```
# ============================ include Firefox profile but nothing else from Mozilla
i:(?i).*\/AppData\/[^\/]+\/Mozilla\/Firefox\/$
i:(?i).*\/AppData\/[^\/]+\/Mozilla\/Firefox\/Profiles\/.*

e:(?i).*\/AppData\/[^\/]+\/Mozilla\/.*
```
the said file was ignored.

So, since i can't check it, i just added it at the bottom, commented, hoping that someone will check if the rule works as intended.

Note: i know it is easy to move all the files from the old profile to the new one, but that takes more steps, plus i think that saving everything under the `/Firefox/` folder feels nicer (no scientific base here, just my whim -> i can be proven wrong).

**Comment by NiJoao** [comment]
**Program files regex:**
Note that
`e:(?i)^[^/]+/Program Files[^/]*/[^/]+/.+`
is an exclusion filter, and in regex101 it seems to do the job, excluding everything except the first subfolder:
https://regex101.com/r/grtjmi/4

But now you gave me the idea of making an inclusion filter instead, and put it before the exclusion of the whole folder, so anyone can uncomment it to get the functionality of storing program names:
`i:(?i)^[^/]+/Program Files[^/]*/[^/]+/$`
What do you think?

**Firefox and Thunderbird**
Thanks!! You are absolutely right.
I checked, Thunderbird also uses a profiles.ini file, so I should be backing up the whole folder.
The only useless folder there in my install is a `Crash Reports` folder, small size.

So I recommend both of these:
```
i:(?i)/AppData/[^/]+/Mozilla/$
i:(?i)/AppData/[^/]+/Mozilla/Firefox/
e:(?i)/AppData/[^/]+/Mozilla/

i:(?i)/AppData/[^/]+/Thunderbird/
```

Also, "not including before confirmation" is a worse policy than including it and exclude/clean later.

**Comment by TowerBR** [comment]
About Thunderbird: you don't need to back up index files (msf files)

I use this: ```e:(?i).*.msf*$```

**Comment by TheBestPessimist** [comment]
@TowerBR thanks, added.

**Comment by TheBestPessimist** [comment]
@NiJoao what do you mean by
> Also, "not including before confirmation" is a worse policy than including it and exclude/clean later.

**Pull request merged**